File: src/nlp.py
"""NLP classifiers: intent, complexity, iteration style with confidence metadata."""

import re
import sqlite3
from collections import Counter

from .db import insert_nlp_enrichments, platform_filter
from .models import Message, Platform

# === Regex Pattern Definitions ===
POLITENESS_PATTERNS = {
    "please": r"\bplease\b",
    "thanks": r"\b(thanks|thank you|thx)\b",
    "sorry": r"\b(sorry|apologies|apologize)\b",
}
BACKTRACK_PATTERNS = {
    "actually": r"\bactually\b",
    "wait": r"\bwait\b",
    "never_mind": r"\b(never\s*mind|nevermind)\b",
    "scratch_that": r"\b(scratch that|ignore that)\b",
}
COMMAND_PATTERN = r"^(please\s+)?(do|make|create|write|build|add|fix|update|change|remove|delete|show|run|help|can you|could you|would you|tell|explain|find|search|get|set|check|test|debug|implement|refactor)\b"
INTENT_PATTERNS = {
    "debug_fix": [
        r"\b(debug|fix|error|bug|failing|broken|traceback|stack trace)\b",
    ],
    "build_feature": [
        r"\b(build|create|implement|add|ship|feature|integrate)\b",
    ],
    "analysis_research": [
        r"\b(analyze|audit|review|compare|benchmark|research|investigate)\b",
    ],
    "explanation_learning": [
        r"\b(explain|why|how does|teach|walk me through|help me understand)\b",
    ],
    "planning_strategy": [
        r"\b(plan|roadmap|milestone|phase|next steps|strategy)\b",
    ],
    "ops_commands": [
        r"\b(run|execute|command|script|deploy|release|push|tag)\b",
    ],
}
ITERATION_MARKERS = [
    r"\bactually\b",
    r"\bwait\b",
    r"\binstead\b",
    r"\bchange\b",
    r"\bupdate\b",
    r"\bfix\b",
    r"\brevise\b",
    r"\bretry\b",
    r"\bagain\b",
    r"\bdifferent\b",
    r"\bscratch that\b",
    r"\brework\b",
]


def classify_intent(text: str) -> tuple[str, float]:
    """Classify a prompt into deterministic intent categories with confidence."""
    scores = {}
    for intent, patterns in INTENT_PATTERNS.items():
        score = 0
        for pattern in patterns:
            if re.search(pattern, text, re.IGNORECASE):
                score += 1
        scores[intent] = score

    top_intent = max(scores.items(), key=lambda item: item[1])[0]
    top_score = scores[top_intent]
    second_score = sorted(scores.values(), reverse=True)[1] if len(scores) > 1 else 0

    if top_score == 0:
        return "other", 0.5

    confidence = 0.6 + min(top_score, 3) * 0.1
    if top_score > second_score:
        confidence += 0.1
    return top_intent, round(min(confidence, 0.95), 2)


def compute_complexity_for_prompt(text: str) -> tuple[float, float]:
    """Compute deterministic complexity score and confidence for a prompt."""
    content = text.strip()
    words = len(content.split())
    score = 1.0

    # Word count contribution
    if words >= 15:
        score += 0.7
    if words >= 35:
        score += 0.8
    if words >= 70:
        score += 0.8

    # Structural and constraint signals
    if "\n" in content:
        score += 0.5
    if content.count(",") >= 2 or ";" in content:
        score += 0.4
    if re.search(r"[{{}}`]|--|\.\/|=|->", content):
        score += 0.5

    constraint_hits = len(
        re.findall(
            r"\b(must|should|without|exactly|at least|at most|step|checklist|constraint)\b",
            content,
            re.IGNORECASE,
        )
    )
    score += min(constraint_hits, 3) * 0.2

    final_score = round(min(score, 5.0), 1)
    confidence = 0.65
    if words >= 20:
        confidence += 0.1
    if constraint_hits > 0:
        confidence += 0.1
    if "\n" in content or re.search(r"[{{}}`]|--|\.\/|=|->", content):
        confidence += 0.05

    return final_score, round(min(confidence, 0.95), 2)


def compute_iteration_style_for_prompt(text: str) -> tuple[float, float]:
    """Compute deterministic iteration-style score and confidence."""
    marker_hits = 0
    for pattern in ITERATION_MARKERS:
        marker_hits += len(re.findall(pattern, text, re.IGNORECASE))

    score = marker_hits * 20
    if "?" in text:
        score += 8
    if re.search(r"\b(again|retry|revise|change|update|different)\b", text, re.IGNORECASE):
        score += 8

    final_score = round(min(score, 100), 1)
    confidence = 0.6 + min(marker_hits, 3) * 0.1
    if "?" in text:
        confidence += 0.05
    return final_score, round(min(confidence, 0.95), 2)


def enrich_nlp(conn: sqlite3.Connection) -> None:
    """Run NLP classifiers on all human messages and store in nlp_enrichments."""
    rows = conn.execute(
        "SELECT id, content FROM messages WHERE role = 'human'"
    ).fetchall()

    enrichments = []
    for msg_id, content in rows:
        intent, intent_conf = classify_intent(content)
        complexity, complexity_conf = compute_complexity_for_prompt(content)
        iteration, iteration_conf = compute_iteration_style_for_prompt(content)
        enrichments.append((
            msg_id, intent, intent_conf, complexity, complexity_conf, iteration, iteration_conf,
        ))

    insert_nlp_enrichments(conn, enrichments)


def compute_nlp_metrics(conn: sqlite3.Connection, platform: Platform | None = None) -> dict:
    """Aggregate NLP enrichments from the database with optional platform filter."""
    pf, pp = platform_filter(platform)

    total = conn.execute(
        f"SELECT COUNT(*) FROM nlp_enrichments e"
        f" JOIN messages m ON e.message_id = m.id"
        f" WHERE m.role = 'human'{pf}",
        pp,
    ).fetchone()[0]

    if total == 0:
        return {
            "intent": {
                "method": "deterministic_rules_v1",
                "counts": {},
                "rates_pct": {},
                "top_intents": [],
                "confidence": {"mean": 0.0, "min": 0.0, "max": 0.0},
            },
            "complexity": {
                "method": "heuristic_complexity_v1",
                "avg_score": 0.0,
                "p50_score": 0.0,
                "p90_score": 0.0,
                "distribution": {"low": 0, "medium": 0, "high": 0},
                "confidence": {"mean": 0.0, "min": 0.0, "max": 0.0},
            },
            "iteration_style": {
                "method": "iteration_markers_v1",
                "avg_score": 0.0,
                "distribution": {"low": 0, "medium": 0, "high": 0},
                "style": "balanced",
                "confidence": {"mean": 0.0, "min": 0.0, "max": 0.0},
            },
        }

    # === Intent ===
    intent_rows = conn.execute(
        f"SELECT e.intent, COUNT(*) as cnt"
        f" FROM nlp_enrichments e JOIN messages m ON e.message_id = m.id"
        f" WHERE m.role = 'human'{pf}"
        f" GROUP BY e.intent ORDER BY cnt DESC, e.intent",
        pp,
    ).fetchall()

    intent_counts = dict(intent_rows)
    intent_rates = {
        intent: round((count / total) * 100, 1)
        for intent, count in intent_rows
    }
    top_intents = [
        {"intent": intent, "count": count, "rate_pct": intent_rates[intent]}
        for intent, count in intent_rows[:3]
    ]

    intent_conf = conn.execute(
        f"SELECT AVG(e.intent_confidence), MIN(e.intent_confidence), MAX(e.intent_confidence)"
        f" FROM nlp_enrichments e JOIN messages m ON e.message_id = m.id"
        f" WHERE m.role = 'human'{pf}",
        pp,
    ).fetchone()

    # === Complexity ===
    complexity_agg = conn.execute(
        f"SELECT AVG(e.complexity_score),"
        f" SUM(CASE WHEN e.complexity_score < 2.5 THEN 1 ELSE 0 END),"
        f" SUM(CASE WHEN e.complexity_score >= 2.5 AND e.complexity_score < 3.8 THEN 1 ELSE 0 END),"
        f" SUM(CASE WHEN e.complexity_score >= 3.8 THEN 1 ELSE 0 END),"
        f" AVG(e.complexity_confidence), MIN(e.complexity_confidence), MAX(e.complexity_confidence)"
        f" FROM nlp_enrichments e JOIN messages m ON e.message_id = m.id"
        f" WHERE m.role = 'human'{pf}",
        pp,
    ).fetchone()

    sorted_complexity = [
        r[0] for r in conn.execute(
            f"SELECT e.complexity_score FROM nlp_enrichments e"
            f" JOIN messages m ON e.message_id = m.id"
            f" WHERE m.role = 'human'{pf}"
            f" ORDER BY e.complexity_score",
            pp,
        ).fetchall()
    ]
    p50_idx = max(0, len(sorted_complexity) // 2 - (1 if len(sorted_complexity) % 2 == 0 else 0))
    p90_idx = max(0, int(len(sorted_complexity) * 0.9) - 1)

    # === Iteration Style ===
    iter_agg = conn.execute(
        f"SELECT AVG(e.iteration_score),"
        f" SUM(CASE WHEN e.iteration_score < 25 THEN 1 ELSE 0 END),"
        f" SUM(CASE WHEN e.iteration_score >= 25 AND e.iteration_score < 60 THEN 1 ELSE 0 END),"
        f" SUM(CASE WHEN e.iteration_score >= 60 THEN 1 ELSE 0 END),"
        f" AVG(e.iteration_confidence), MIN(e.iteration_confidence), MAX(e.iteration_confidence)"
        f" FROM nlp_enrichments e JOIN messages m ON e.message_id = m.id"
        f" WHERE m.role = 'human'{pf}",
        pp,
    ).fetchone()

    avg_iteration = round(iter_agg[0], 1)
    if avg_iteration >= 60:
        iteration_style = "highly_iterative"
    elif avg_iteration >= 30:
        iteration_style = "balanced_iterative"
    else:
        iteration_style = "direct"

    return {
        "intent": {
            "method": "deterministic_rules_v1",
            "counts": intent_counts,
            "rates_pct": intent_rates,
            "top_intents": top_intents,
            "confidence": {
                "mean": round(intent_conf[0], 2),
                "min": round(intent_conf[1], 2),
                "max": round(intent_conf[2], 2),
            },
        },
        "complexity": {
            "method": "heuristic_complexity_v1",
            "avg_score": round(complexity_agg[0], 1),
            "p50_score": round(sorted_complexity[p50_idx], 1),
            "p90_score": round(sorted_complexity[p90_idx], 1),
            "distribution": {
                "low": complexity_agg[1],
                "medium": complexity_agg[2],
                "high": complexity_agg[3],
            },
            "confidence": {
                "mean": round(complexity_agg[4], 2),
                "min": round(complexity_agg[5], 2),
                "max": round(complexity_agg[6], 2),
            },
        },
        "iteration_style": {
            "method": "iteration_markers_v1",
            "avg_score": avg_iteration,
            "distribution": {
                "low": iter_agg[1],
                "medium": iter_agg[2],
                "high": iter_agg[3],
            },
            "style": iteration_style,
            "confidence": {
                "mean": round(iter_agg[4], 2),
                "min": round(iter_agg[5], 2),
                "max": round(iter_agg[6], 2),
            },
        },
    }

File: src/pipeline/nlp.ts
import type { Client } from "@libsql/client";
import { insertNlpEnrichments, platformFilter, type MessageRow } from "./db.js";
import type { Platform } from "./models.js";

export const POLITENESS_PATTERNS: Record<string, RegExp> = {
  please: /\bplease\b/gi,
  thanks: /\b(thanks|thank you|thx)\b/gi,
  sorry: /\b(sorry|apologies|apologize)\b/gi,
};

export const BACKTRACK_PATTERNS: Record<string, RegExp> = {
  actually: /\bactually\b/gi,
  wait: /\bwait\b/gi,
  never_mind: /\b(never\s*mind|nevermind)\b/gi,
  scratch_that: /\b(scratch that|ignore that)\b/gi,
};

export const COMMAND_PATTERN =
  /^(please\s+)?(do|make|create|write|build|add|fix|update|change|remove|delete|show|run|help|can you|could you|would you|tell|explain|find|search|get|set|check|test|debug|implement|refactor)\b/i;

export const INTENT_PATTERNS: Record<string, RegExp[]> = {
  debug_fix: [/\b(debug|fix|error|bug|failing|broken|traceback|stack trace)\b/i],
  build_feature: [/\b(build|create|implement|add|ship|feature|integrate)\b/i],
  analysis_research: [/\b(analyze|audit|review|compare|benchmark|research|investigate)\b/i],
  explanation_learning: [/\b(explain|why|how does|teach|walk me through|help me understand)\b/i],
  planning_strategy: [/\b(plan|roadmap|milestone|phase|next steps|strategy)\b/i],
  ops_commands: [/\b(run|execute|command|script|deploy|release|push|tag)\b/i],
};

const ITERATION_MARKERS: RegExp[] = [
  /\bactually\b/gi,
  /\bwait\b/gi,
  /\binstead\b/gi,
  /\bchange\b/gi,
  /\bupdate\b/gi,
  /\bfix\b/gi,
  /\brevise\b/gi,
  /\bretry\b/gi,
  /\bagain\b/gi,
  /\bdifferent\b/gi,
  /\bscratch that\b/gi,
  /\brework\b/gi,
];

export function classifyIntent(text: string): { intent: string; confidence: number } {
  const scores: Record<string, number> = {};
  for (const [intent, patterns] of Object.entries(INTENT_PATTERNS)) {
    let score = 0;
    for (const pattern of patterns) {
      if (pattern.test(text)) score++;
      pattern.lastIndex = 0;
    }
    scores[intent] = score;
  }

  const entries = Object.entries(scores).sort((a, b) => b[1] - a[1]);
  const topIntent = entries[0][0];
  const topScore = entries[0][1];
  const secondScore = entries.length > 1 ? entries[1][1] : 0;

  if (topScore === 0) return { intent: "other", confidence: 0.5 };

  let confidence = 0.6 + Math.min(topScore, 3) * 0.1;
  if (topScore > secondScore) confidence += 0.1;
  return { intent: topIntent, confidence: round(Math.min(confidence, 0.95), 2) };
}

export function computeComplexity(text: string): { score: number; confidence: number } {
  const content = text.trim();
  const words = content.split(/\s+/).filter(Boolean).length;
  let score = 1.0;

  if (words >= 15) score += 0.7;
  if (words >= 35) score += 0.8;
  if (words >= 70) score += 0.8;

  if (content.includes("\n")) score += 0.5;
  if ((content.match(/,/g)?.length ?? 0) >= 2 || content.includes(";")) score += 0.4;
  if (/[{}`]|--|\.\/|=|->/.test(content)) score += 0.5;

  const constraintHits = (
    content.match(/\b(must|should|without|exactly|at least|at most|step|checklist|constraint)\b/gi) ?? []
  ).length;
  score += Math.min(constraintHits, 3) * 0.2;

  const finalScore = round(Math.min(score, 5.0), 1);
  let confidence = 0.65;
  if (words >= 20) confidence += 0.1;
  if (constraintHits > 0) confidence += 0.1;
  if (content.includes("\n") || /[{}`]|--|\.\/|=|->/.test(content)) confidence += 0.05;

  return { score: finalScore, confidence: round(Math.min(confidence, 0.95), 2) };
}

export function computeIterationStyle(text: string): { score: number; confidence: number } {
  let markerHits = 0;
  for (const pattern of ITERATION_MARKERS) {
    markerHits += (text.match(pattern) ?? []).length;
    pattern.lastIndex = 0;
  }

  let score = markerHits * 20;
  if (text.includes("?")) score += 8;
  if (/\b(again|retry|revise|change|update|different)\b/i.test(text)) score += 8;

  const finalScore = round(Math.min(score, 100), 1);
  let confidence = 0.6 + Math.min(markerHits, 3) * 0.1;
  if (text.includes("?")) confidence += 0.05;

  return { score: finalScore, confidence: round(Math.min(confidence, 0.95), 2) };
}

export async function enrichNlp(client: Client): Promise<number> {
  const result = await client.execute(
    "SELECT m.id, m.content FROM messages m LEFT JOIN nlp_enrichments e ON m.id = e.message_id WHERE m.role = 'human' AND m.is_excluded = 0 AND e.message_id IS NULL",
  );

  const enrichments = result.rows.map((row) => {
    const content = String(row.content);
    const intent = classifyIntent(content);
    const complexity = computeComplexity(content);
    const iteration = computeIterationStyle(content);
    return {
      messageId: Number(row.id),
      intent: intent.intent,
      intentConfidence: intent.confidence,
      complexityScore: complexity.score,
      complexityConfidence: complexity.confidence,
      iterationScore: iteration.score,
      iterationConfidence: iteration.confidence,
    };
  });

  if (enrichments.length > 0) {
    await insertNlpEnrichments(client, enrichments);
  }
  return enrichments.length;
}

async function aggregateClassifier(
  client: Client,
  column: string,
  pf: { clause: string; args: any[] },
): Promise<{ avg_score: number | null; confidence: { mean: number; min: number; max: number } | null }> {
  const result = await client.execute({
    sql: `SELECT AVG(e.${column}) as avg_s, AVG(e.${column.replace("_score", "_confidence")}) as avg_c, MIN(e.${column.replace("_score", "_confidence")}) as min_c, MAX(e.${column.replace("_score", "_confidence")}) as max_c FROM nlp_enrichments e JOIN messages m ON e.message_id = m.id WHERE m.role = 'human' AND m.is_excluded = 0 AND e.${column} IS NOT NULL${pf.clause}`,
    args: pf.args,
  });
  const r = result.rows[0];
  if (r.avg_s == null) return { avg_score: null, confidence: null };
  return {
    avg_score: round(Number(r.avg_s), 1),
    confidence: { mean: round(Number(r.avg_c), 2), min: round(Number(r.min_c), 2), max: round(Number(r.max_c), 2) },
  };
}

export async function computeNlpMetrics(
  client: Client,
  platform?: Platform,
): Promise<Record<string, any>> {
  const pf = platformFilter(platform);

  const totalResult = await client.execute({
    sql: `SELECT COUNT(*) as cnt FROM nlp_enrichments e JOIN messages m ON e.message_id = m.id WHERE m.role = 'human' AND m.is_excluded = 0${pf.clause}`,
    args: pf.args,
  });
  const total = Number(totalResult.rows[0].cnt);

  const emptyClassifier = { method: "embedding_similarity_v1", avg_score: null, confidence: null };

  if (total === 0) {
    return {
      intent: { method: "deterministic_rules_v1", counts: {}, rates_pct: {}, top_intents: [], confidence: { mean: 0, min: 0, max: 0 } },
      hitl_score: emptyClassifier,
      vibe_coder_index: emptyClassifier,
      politeness: emptyClassifier,
    };
  }

  // Intent
  const intentRows = await client.execute({
    sql: `SELECT e.intent, COUNT(*) as cnt FROM nlp_enrichments e JOIN messages m ON e.message_id = m.id WHERE m.role = 'human' AND m.is_excluded = 0${pf.clause} GROUP BY e.intent ORDER BY cnt DESC, e.intent`,
    args: pf.args,
  });
  const intentCounts: Record<string, number> = {};
  const intentRates: Record<string, number> = {};
  const topIntents: any[] = [];
  for (const row of intentRows.rows) {
    const intent = String(row.intent);
    const count = Number(row.cnt);
    intentCounts[intent] = count;
    intentRates[intent] = round((count / total) * 100, 1);
  }
  for (const row of intentRows.rows.slice(0, 3)) {
    const intent = String(row.intent);
    topIntents.push({ intent, count: intentCounts[intent], rate_pct: intentRates[intent] });
  }

  const intentConf = await client.execute({
    sql: `SELECT AVG(e.intent_confidence) as avg_c, MIN(e.intent_confidence) as min_c, MAX(e.intent_confidence) as max_c FROM nlp_enrichments e JOIN messages m ON e.message_id = m.id WHERE m.role = 'human' AND m.is_excluded = 0${pf.clause}`,
    args: pf.args,
  });
  const ic = intentConf.rows[0];

  // Aggregate hero embedding classifiers in parallel
  const [hitl, vibe, politeness] = await Promise.all([
    aggregateClassifier(client, "hitl_score", pf),
    aggregateClassifier(client, "vibe_score", pf),
    aggregateClassifier(client, "politeness_score", pf),
  ]);

  return {
    intent: {
      method: "deterministic_rules_v1",
      counts: intentCounts,
      rates_pct: intentRates,
      top_intents: topIntents,
      confidence: { mean: round(Number(ic.avg_c), 2), min: round(Number(ic.min_c), 2), max: round(Number(ic.max_c), 2) },
    },
    hitl_score: { method: "embedding_similarity_v1", ...hitl },
    vibe_coder_index: { method: "embedding_similarity_v1", ...vibe },
    politeness: { method: "embedding_similarity_v1", ...politeness },
  };
}

function round(n: number, decimals: number): number {
  const factor = 10 ** decimals;
  return Math.round(n * factor) / factor;
}
